instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go

```diff
@@ -0,0 +1,351 @@
+package otellambda
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"reflect"
+	"runtime"
+	"strings"
+
+	"github.com/aws/aws-lambda-go/lambda"
+	"github.com/aws/aws-lambda-go/lambdacontext"
+
+	lambdadetector "go.opentelemetry.io/contrib/detectors/aws/lambda"
+	"go.opentelemetry.io/contrib/propagators/aws/xray"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	tracerName = "go.opentelemetry.io/contrib/instrumentation/github.com/aws/aws-lambda-go/otellambda"
+)
+
+var tp *sdktrace.TracerProvider
+var errorLogger = log.New(log.Writer(), "OTel Lambda Error: ", 0)
+
+func init() {
+	otel.SetTextMapPropagator(xray.Propagator{})
+}
+
+func initTracerProvider() {
+	ctx := context.Background()
+
+	exp, err := otlptracegrpc.New(ctx, otlptracegrpc.WithInsecure())
+	if err != nil {
+		errorLogger.Printf("failed to initialize exporter: %v\n", err)
+		return
+	}
+
+	detector := lambdadetector.NewResourceDetector()
+	res, err := detector.Detect(ctx)
+	if err != nil {
+		errorLogger.Printf("failed to detect lambda resources: %v\n", err)
+		return
+	}
+
+	tp = sdktrace.NewTracerProvider(
+		sdktrace.WithBatcher(exp),
+		sdktrace.WithIDGenerator(xray.NewIDGenerator()),
+		sdktrace.WithResource(res),
+	)
+
+	// Set the traceprovider
+	otel.SetTracerProvider(tp)
+}
+
+func errorHandler(e error) func(context.Context, interface{}) (interface{}, error) {
+	return func(context.Context, interface{}) (interface{}, error) {
+		return nil, e
+	}
+}
+
+// Ensure handler takes 0-2 values, with context
+// as its first value if two arguments exist
+func validateArguments(handler reflect.Type) (bool, error) {
+	handlerTakesContext := false
+	if handler.NumIn() > 2 {
+		return false, fmt.Errorf("handlers may not take more than two arguments, but handler takes %d", handler.NumIn())
+	} else if handler.NumIn() > 0 {
```

Comment:
Perhaps use a `switch handler.NumIn()` here, and add a `case 1` that avoids doing reflection on the only argument when if a context can only be provided with two arguments.

Reply:
Context can be the only argument and when we are only given one argument we still need to know whether it is context or an event in order to correctly handle an event in the LambdaHandlerWrapper so I believe the reflection is necessary either way.